options/internal/generic/essential.cpp

```diff
@@ -189,14 +189,21 @@ void *memset(void *dest, int byte, size_t count) {
 // --------------------------------------------------------------------------------------
 
 void *memmove(void *dest, const void *src, size_t size) {
-	char *dest_bytes = (char *)dest;
-	char *src_bytes = (char *)src;
-	if(dest_bytes < src_bytes) {
+	// Use uintptr_t for pointer comparisons because otherwise it's undefined behaviour
+	// when dest and src point to different objects.
+	uintptr_t udest = (uintptr_t)dest;
+	uintptr_t usrc = (uintptr_t)src;
+
+	if(udest < usrc || usrc + size <= udest) {
```

Comment:
Is there any reason why this would only to memmove in the `usrc < udest` case?

Reply:
forward_copy will clobber itself when src < dst and they overlap

Reply:
This comment was probably about the case that `src < dst` and there is no overlap.

Reply:
Don't quite understand -- that's exactly the case I'm handling via memcpy now

Reply:
Ah I misread the condition.